### PR TITLE
Fade map when time range picker is open

### PR DESCRIPTION
**Pull request by 3v0k4**
^^^^^^^^

### CODE REVIEW

File: app/javascript/elm/src/Main.elm [hunk 1]
```diff
@@ -509,7 +530,32 @@ update msg model =
...
-            ( { model | isShowingTimeRangeFilter = isShown }, Cmd.none )
+            let
+                overlay =
+                    if isShown then
+                        Overlay
+
+                    else
+                        NoOverlay
+            in
+            ( { model | overlay = updateOverlay overlay model }, Cmd.none )
```
Comment (AkkarinForest) [inline app/javascript/elm/src/Main.elm]: What do you think about having just one generic update case like so:
```
        ChangeOverlay overlayType onOrOff ->
            let
                overlay =
                    if onOrOff then
                        overlayType

                    else
                        NoOverlay
            in
            ( { model | overlay = overlay }, Cmd.none )

```

Then you can call it like so 

```
, Ports.updateIsHttping (ChangeOverlay OverlayWithSpinner)
, Ports.isShowingTimeRangeFilter (ChangeOverlay BasicOverlay)
```
or
```
        ClosePopup ->
            ( { model | popup = Popup.None, overlay = NoOverlay }, Cmd.none )

```

That would let us remove the duplication from update branches and make the logic reusable.

File: app/javascript/elm/src/Main.elm [hunk 2]
```diff
@@ -349,7 +362,15 @@ update msg model =
...
-            ( { model | isHttping = isHttpingNow }, Cmd.none )
+            let
+                overlay =
+                    if isHttpingNow then
+                        HttpingOverlay
+
+                    else
+                        NoOverlay
+            in
+            ( { model | isHttping = isHttpingNow, overlay = updateOverlay overlay model }, Cmd.none )
```
Comment (AkkarinForest) [inline app/javascript/elm/src/Main.elm]: I guess no need to keep `isHttping` in the model anymore?

**Comment by 3v0k4** [comment]
As discussed with @AkkarinForest offline we couldn't address his comments w/o a heavy refactoring. The latest state of the CR should have addressed them

File: app/javascript/elm/src/Data/Overlay.elm [hunk 3]
```diff
@@ -0,0 +1,85 @@
+module Data.Overlay exposing (Model, Operation(..), Overlay(..), init, none, update, view)
+
+import Html exposing (Html, div, p, text)
+import Html.Attributes exposing (class, id)
+import List.Nonempty as Nonempty exposing (Nonempty(..))
+
+
+type Model
+    = Overlays (Nonempty Overlay)
+    | NoOverlay
```
Comment (AkkarinForest) [inline app/javascript/elm/src/Data/Overlay.elm]: So this is equivalent to `type Model = Overlays (List Overlay)` where `NoOverlay` covers the case of an empty list, right? Can you share what are the advantages of that?
Cause I can see that having a normal list would make the updated shorter - you would only need @ branches - for adding and filtering the list. 
And in the view, you can just have `case (head model) of`.

Comment (3v0k4) [reply app/javascript/elm/src/Data/Overlay.elm]: Good point. Removed nonempty but kept the opaque type. Thanks!

File: app/javascript/elm/src/Data/Overlay.elm [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        ( RemoveOverlay overlay, Overlays overlays ) ->
+            Overlays <| List.filter ((/=) overlay) overlays
+
+
+view : Model -> Html msg
+view (Overlays overlays) =
+    case List.head overlays of
+        Just overlay ->
+            case overlay of
+                HttpingOverlay ->
```
Comment (gitjul) [inline app/javascript/elm/src/Data/Overlay.elm]: I'd probably go with `Loading`

**Comment by gitjul** [comment]
👍 👍 👏

File: app/javascript/elm/src/Data/Overlay.elm (hunk 4 above)
Comment (3v0k4) [reply app/javascript/elm/src/Data/Overlay.elm]: I like your suggestion better but for now I'm leaving it the way it is cause the verb httping is used in multiple places and prefer consistency now. I'm not sure how I would rename the rest

**Comment by 3v0k4** [comment]
@gitjul thanks for the CR!

**Pull request merged**